dace/transformation/interstate/state_fusion.py

```diff
@@ -125,38 +210,122 @@ def can_be_applied(graph, candidate, expr_index, sdfg, strict=False):
             # Apply transformation in case all paths to the second state's
             # nodes go through the same access node, which implies sequential
             # behavior in SDFG semantics.
-            check_strict = len(first_cc)
-            for cc_output in first_cc_output:
-                out_nodes = [
-                    n for n in first_state.sink_nodes() if n in cc_output
-                ]
-                # Branching exists, multiple paths may involve same access node
-                # potentially causing data races
-                if len(out_nodes) > 1:
-                    continue
-
-                # Otherwise, check if any of the second state's connected
-                # components for matching input
-                for node in out_nodes:
-                    if (next((x for x in second_input if x.label == node.label),
-                             None) is not None):
-                        check_strict -= 1
-                        break
-
-            if check_strict > 0:
-                # Check strict conditions
-                # RW dependency
-                for node in first_input:
-                    if (next(
-                        (x for x in second_output if x.label == node.label),
-                            None) is not None):
-                        return False
-                # WW dependency
-                for node in first_output:
-                    if (next(
-                        (x for x in second_output if x.label == node.label),
-                            None) is not None):
-                        return False
+            first_output_names = {node.data for node in first_output}
+            second_input_names = {node.data for node in second_input}
+
+            # If any second input appears more than once, fail
```

Comment:
My understanding is that this check and the next one are supposed to make the transformation fail because of the difficulty of handling such cases, and not because the states cannot be fused. If we merged the nodes that appear twice, then we should be able to proceed with the transformation. Is this correct?

Reply:
Exactly. There is another strict transformation called `MergeInputs` (or something similar) that takes care of this situation.